pallets/creator-staking/src/functions.rs

```diff
@@ -0,0 +1,462 @@
+            .map_err(|_| Error::<T>::CannotChangeStakeInPastEra)?;
+
+        Self::ensure_max_era_stake_items_not_exceeded(backer_info)?;
+
+        ensure!(
+            backer_info.current_stake() >= T::MinimumStake::get(),
+            Error::<T>::InsufficientStakingAmount,
+        );
+
+        // Increment total backer's deposit for creator.
+        staking_info.total = staking_info.total.saturating_add(desired_amount);
+
+        Ok(())
+    }
+
+    /// An utility method used to unstake specified amount from an arbitrary creator.
+    ///
+    /// The amount unstaked can be different in case staked amount would fall bellow
+    /// `MinimumStakingAmount`. In that case, entire staked amount will be unstaked.
+    ///
+    /// `BackerInfo` and `CreatorStakeInfo` are provided and all checks are made to ensure that
+    /// it's possible to complete unstake operation.
+    ///
+    /// # Arguments
+    ///
+    /// * `backer_info` - info about backer's stakes on the creator up to current moment
+    /// * `staking_info` - general info about creator stakes up to current moment
+    /// * `value` - value which should be unstaked
+    /// * `current_era` - current creators-staking era
+    ///
+    /// # Returns
+    ///
+    /// If unstake operation was successful, given structs are properly modified and total
+    /// unstaked value is returned. If not, an error is returned and structs are left in
+    /// an undefined state.
+    pub(super) fn calculate_final_unstaking_amount(
+        backer_info: &mut StakesInfoOf<T>,
+        stake_info: &mut CreatorStakeInfo<BalanceOf<T>>,
+        desired_amount: BalanceOf<T>,
+        current_era: EraIndex,
+    ) -> Result<BalanceOf<T>, DispatchError> {
+        let staked_value = backer_info.current_stake();
+        ensure!(staked_value > Zero::zero(), Error::<T>::NotStakedCreator);
+
+        // Calculate the value which will be unstaked.
+        let remaining = staked_value.saturating_sub(desired_amount);
+
+        // If remaining amount is less than minimum staking amount, unstake the entire amount.
+        let amount_to_unstake = if remaining < T::MinimumStake::get() {
+            stake_info.backers_count = stake_info.backers_count.saturating_sub(1);
+            staked_value
+        } else {
+            desired_amount
+        };
+
+        // Sanity check
+        ensure!(amount_to_unstake > Zero::zero(), Error::<T>::CannotUnstakeZero);
+
+        stake_info.total = stake_info.total.saturating_sub(amount_to_unstake);
+
+        backer_info
+            .unstake(current_era, amount_to_unstake)
+            .map_err(|_| Error::<T>::CannotChangeStakeInPastEra)?;
+
+        Self::ensure_max_era_stake_items_not_exceeded(backer_info)?;
+
+        Ok(amount_to_unstake)
+    }
+
+    /// Update the locks for a backer. This will also update the stash lock.
+    /// This lock will lock the entire funds except paying for further transactions.
```

Comment:
Thank you for mentioning this. For now I do `WithdrawReasons` to be `TRANSACTION_PAYMENT` and `RESERVE`, which means you'll be able to spend locked funds in order to pay for (system) transaction costs and in order to reserve some funds for a later return or repatriation.

Reply:
So I can lock 100% of my tokens, and then use those to pay for tx fees, and they'll be instantly unlocked and burnt/sent to the treasury?

Reply:
If you still have unlocked tokens (because of `MinimumRemainingFreeBalance`), then what's the point of this part of the code? Why does being able to use locked tokens to pay for tx matter if you still have unlocked tokens?

Reply:
> If you still have unlocked tokens (because of `MinimumRemainingFreeBalance`), then what's the point of this part of the code? Why does being able to use locked tokens to pay for tx matter if you still have unlocked tokens?

`update_backer_locks` functions (this part of code) is responsible for actually locking tokens, not to allow transferring them (for example). However, you can still reserve them in any other pallet, because reserved tokens are usually meant to be returned if you, for example, free a storage.

Though, maybe it's not actually good idea to allow withdrawing lock at all.

The `MinimumRemainingFreeBalance` is to leave you some free balance, because in other way your account may be killed. After that, you'll still be able to claim rewards (paying fees from your lock), but their amount may be less than existential deposit, so they will probably be burnt immediately.